### ISSUE TITLE
Error connecting two services to a third

### ISSUE THREAD

**Issue by tombeynon** [ovrclk/akash#1303]
Connecting a single service to another with `expose.to` works great. Introducing a second service connected in the same way causes a `provider error messsage: kube: no global services for lease` error message and the two dependent containers don't start. Removing the `expose.to` instructions for one of the containers will then fix the error, and (incorrectly?) both containers do have access to the third. This functions as a workaround.

The below SDL works until you uncomment `- service: test-2`. I tried a few variations of this (duplicating the whole `port` section etc) with no luck.

```
---
version: "2.0"

services:
  test-1:
    image: tombeynon/hello-world-curl-test
    depends_on:
      - service: hello-world
  test-2:
    image: tombeynon/hello-world-curl-test
    depends_on:
      - service: hello-world
  hello-world:
    image: tombeynon/akash-hello-world:release-v0.1.1
    expose:
      - port: 80
        to:
          - service: test-1
          # - service: test-2
      - port: 9999 # required global service
        to:
          - global: true

profiles:
  compute:
    test-1:
      resources:
        cpu:
          units: 0.1
        memory:
          size: 0.1Gi
        storage:
          size: 0.1Gi
    test-2:
      resources:
        cpu:
          units: 0.1
        memory:
          size: 0.1Gi
        storage:
          size: 0.1Gi
    hello-world:
      resources:
        cpu:
          units: 0.1
        memory:
          size: 0.1Gi
        storage:
          size: 0.1Gi
  placement:
    dcloud:
      pricing:
        test-1:
          denom: uakt
          amount: 10
        test-2:
          denom: uakt
          amount: 10
        hello-world:
          denom: uakt
          amount: 10

deployment:
  test-1:
    dcloud:
      profile: test-1
      count: 1
  test-2:
    dcloud:
      profile: test-2
      count: 1
  hello-world:
    dcloud:
      profile: hello-world
      count: 1
```

**Comment by hydrogen18** [comment]
My provider spit this out when I tried your SDL

```
I[2021-07-09|14:03:40.420] manifest received                            module=provider-cluster cmp=service
I[2021-07-09|14:03:40.434] No deployments found for                     module=provider-cluster-kube leasenamespace=4homp4j2rl698e0k9f65boov74ujgsotmr5uks5qp33gs
D[2021-07-09|14:03:40.447] provider/cluster/kube/builder: created service module=kube-builder service="&Service{ObjectMeta:{hello-world      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[akash.network:true akash.network/manifest-service:hello-world akash.network/namespace:4homp4j2rl698e0k9f65boov74ujgsotmr5uks5qp33gs] map[] [] []  []},Spec:ServiceSpec{Ports:[]ServicePort{ServicePort{Name:1-80,Protocol:TCP,Port:80,TargetPort:{0 80 },NodePort:0,AppProtocol:nil,},ServicePort{Name:2-80,Protocol:TCP,Port:80,TargetPort:{0 80 },NodePort:0,AppProtocol:nil,},},Selector:map[string]string{akash.network: true,akash.network/manifest-service: hello-world,akash.network/namespace: 4homp4j2rl698e0k9f65boov74ujgsotmr5uks5qp33gs,},ClusterIP:,Type:ClusterIP,ExternalIPs:[],SessionAffinity:,LoadBalancerIP:,LoadBalancerSourceRanges:[],ExternalName:,ExternalTrafficPolicy:,HealthCheckNodePort:0,PublishNotReadyAddresses:false,SessionAffinityConfig:nil,IPFamily:nil,TopologyKeys:[],},Status:ServiceStatus{LoadBalancer:LoadBalancerStatus{Ingress:[]LoadBalancerIngress{},},},}"
E[2021-07-09|14:03:40.448] applying local service                       module=provider-cluster-kube err="Service \"hello-world\" is invalid: spec.ports[1]: Duplicate value: core.ServicePort{Name:\"\", Protocol:\"TCP\", AppProtocol:(*string)(nil), Port:80, TargetPort:intstr.IntOrString{Type:0, IntVal:0, StrVal:\"\"}, NodePort:0}" lease=akash1lz5dvghpxy0uv2fvags00d5wr9h8kka8g3hekj/1002/1/1/akash1uvj4vc3m7pgh5n8jezev08y96nhmgap73z7kqv service=hello-world
E[2021-07-09|14:03:40.448] execution error                              module=provider-cluster cmp=service cmp=deployment-manager lease=akash1lz5dvghpxy0uv2fvags00d5wr9h8kka8g3hekj/1002/1/1/akash1uvj4vc3m7pgh5n8jezev08y96nhmgap73z7kqv manifest-group=dcloud state=deploy-active err="Service \"hello-world\" is invalid: spec.ports[1]: Duplicate value: core.ServicePort{Name:\"\", Protocol:\"TCP\", AppProtocol:(*string)(nil), Port:80, TargetPort:intstr.IntOrString{Type:0, IntVal:0, StrVal:\"\"}, NodePort:0}"

```

unsure what the cause is exactly but the issue is legitimate.

**Comment by boz** [comment]
```yaml
    depends_on:
      - service: hello-world
  test-2:
    image: tombeynon/hello-world-curl-test
    depends_on:
      - service: hello-world
```

`depends_on` is a no-op right now.

**Comment by tombeynon** [comment]
> `depends_on` is (or should be) a no-op right now.

That makes sense actually, it didn't seem to have much effect. I included it for completeness but with or without doesn't change the behaviour here.

**Comment by hydrogen18** [comment]
The provider attempted to create two service ports in kube for the same port on the same service. This obviously wouldn't make any sense as you can't use the same port twice in a service (it's an IP address, basically).

This happens because the SDL file interprets each entry in the `to:` list in the YAML as its own service expose definition, so specify a list with a length greater than one always duplicates the port. This is probably a bug in the way we are creating our internal representation of a manifest.

This happens here : https://github.com/ovrclk/akash/blob/04e7a7667dd94b5a15fa039e4f7df5c9ad93be4f/sdl/v2.go#L133 

The external port should be at a higher level, since it is necessarily duplicated when the list has a length greater than one (e.g. in Tom's example). A single ServicePort definition should be enough to make things work in Kubernetes.

**Comment by hydrogen18** [comment]
Hi @tombeynon ,

I've made some changes in my local test environment and the provider now deploys your SDL.

Is the container you have supposed to exit immediately? It logs this 

```
│   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current                                                                           │
│                                  Dload  Upload   Total   Spent    Left  Speed                                                                             │
│ HTTP/1.1 200 OK                                                                                                                                           │
│   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0  0  1297    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--    │
│ Server: nginx/1.19.8                                                                                                                                      │
│ Date: Wed, 14 Jul 2021 19:03:46 GMT                                                                                                                       │
│ Content-Type: text/html                                                                                                                                   │
│ Content-Length: 1297                                                                                                                                      │
│ Last-Modified: Mon, 15 Mar 2021 12:21:20 GMT                                                                                                              │
│ Connection: keep-alive                                                                                                                                    │
│ ETag: "604f5140-511"                                                                                                                                      │
│ Accept-Ranges: bytes                                                                                                                                      │
│ stream closed                                                                                                                                             │
│                                                        
```

**Comment by tombeynon** [comment]
Hey @hydrogen18 - yes that's correct. The `hello-world-curl-test` containers just curl the `hello-world` container then exit. If the bug persists, one or both of the containers won't be able to connect. Otherwise they both can.

**Comment by hydrogen18** [comment]
OK great, that confirms my fix is working. Thanks!

**Comment by coffeeroaster** [comment]
I'm late to the party, but here's an example of 4 services talking to each other: https://github.com/coffeeroaster/unstoppable-web2.0/blob/main/akash/deploy-sample.yml

**Comment by hydrogen18** [comment]
This should be fixed in v0.14.0, can someone confirm please?

**Comment by hydrogen18** [comment]
Closing since this should be fixed & there has been no other feedback
